ADR/0032-decoupling-deployment.md

```diff
@@ -0,0 +1,201 @@
+# 32. Decoupling Deployment
+
+Date: 2023-11-17
+
+## Status
+
+Draft
+
+## Authors
+
+- Ralph Bean
+
+## Context
+
+Since the beginning of our project, we've had an emphasis on providing an integrated experience for
+the user, automating all steps through build, test, deployment, and release to higher environments.
+
+Some challenges:
+
+- Our controllers' APIs are highly coupled. For example, it isn't feasible to try and use
+  [integration-service] today without [application-service] and [gitops-service]. This creates
+  a high barrier to engage with some of the useful subsystems we've built. A user (and potential
+  contributor) has to adopt the whole thing, or nothing.
+- Today, we can generate simple but often incorrect deployment manifests for users' applications,
+  which we deploy via the application gitops repo. The user cannot interact with or influence these
+  resources directly; they have to do so through the Application/Component API. If we permit them to
+  directly provide deployment resources through that API, the API will become a leaky abstraction.
+  If don't permit them to provide their own resources, and instead try to handle every application
+  configuration case in the HAS generation code, we will struggle to keep up with users' demands for
+  the many and varied kinds of application deployments.
+
+See also [RHTAP-1873](https://issues.redhat.com/browse/RHTAP-1873).
+
+## Decision
+
+We are going to decouple the deployment from the build, test, and release portions of our system.
+
+**Deployment**:
+
+- The [Environment], [SnapshotEnvironmentBinding], and [GitOpsDeploymentManagedEnvironment]
+  resources will be deprecated and eventually decomissioned.
+- The [application-service] will stop generating GitOps repo content.
+- HAC will stop rendering [Environments] and their status.
+- If a user wants to make use of deployment capabilities, we will promote the usage of [renovatebot]
+  to propagate _released_ images to their self-managed gitops repo as pull requests.
```

Comment:
What might this process look like? What type of repository would Renovate be running on for ensuring up-to-date deployment capabilities? The latency for depending on Renovate seems like it would be too high to be able to run any kind of integration test in succession with the Snapshot generation, therefore causing additional challenges for quickly testing and releasing Snapshots.

Additionally, if the system doesn't know how to deploy components, how would PR-level deployments/integration tests be triggered? Would there be different locations for Renovate-driven updates for PRs as push events?

Reply:
After reading further, I assume that my reading _here_ was misdirected. The proposal to use Renovate is _only_ for deploying to specific environments. There would need to be configurations present in whatever deployment strategy/repo (i.e. Argo/GitOps) which contains pullspecs that can be updated by Renovate in order to deploy to a long-running environment.

Renovate is not proposed for the flow including the testing of specific built artifacts within the pipeline/process.

Reply:
Yes, right!